File: samples/cext-trey-research-auth/src/services/ProjectApiService.ts
import { Project, HoursEntry, Assignment } from '../model/baseModel';
import { ApiProject, ApiAddConsultantToProjectResponse } from '../model/apiModel';
import ProjectDbService from './ProjectDbService';
import AssignmentDbService from './AssignmentDbService';
import ConsultantDbService from './ConsultantDbService';
import ConsultantApiService from './ConsultantApiService';
import { HttpError } from './Utilities';

class ProjectApiService {

    async getApiProjectById(projectId: string): Promise<ApiProject> {
        const project = await ProjectDbService.getProjectById(projectId);
        let assignments = await AssignmentDbService.getAssignments();

        const result = await this.getApiProject(project, assignments);
        return result;
    }

    async getApiProjects(projectOrClientName: string, consultantName: string): Promise<ApiProject[]> {

        let projects = await ProjectDbService.getProjects();
        let assignments = await AssignmentDbService.getAssignments();

        // Filter on base properties
        if (projectOrClientName) {
            projects = projects.filter(
                (p) => {
                    const name = p.name?.toLowerCase();
                    const clientName = p.clientName?.toLowerCase();
                    return name.includes(projectOrClientName.toLowerCase()) || clientName.includes(projectOrClientName.toLowerCase());
                });
        }
        //remove duplicates
        projects = projects.filter(
            (project, index, self) => 
                index === self.findIndex((p) => (
                    p.id === project.id
                ))
        );

        // Augment the base properties with assignment information
        let result = await Promise.all(projects.map((p) => this.getApiProject(p, assignments)));

        // Filter on augmented properties
        if (result && consultantName) {
            result = result.filter(
                (p) => {
                    const name = consultantName.toLowerCase();
                    return p.consultants.find((n) => n.consultantName.toLowerCase().includes(name));
                });
        };

        return result;
    }

    // Augment a project to get an ApiProject
    async getApiProject(project: Project, assignments: Assignment[]): Promise<ApiProject> {

        const result = project as ApiProject;
        assignments = assignments.filter((a) => a.projectId === project.id);

        result.consultants = [];
        result.forecastThisMonth = 0;
        result.forecastNextMonth = 0;
        result.deliveredLastMonth = 0;
        result.deliveredThisMonth = 0;

        for (let assignment of assignments) {
            const consultant = await ConsultantDbService.getConsultantById(assignment.consultantId);
            const { lastMonthHours: forecastLastMonth,
                thisMonthHours: forecastThisMonth,
                nextMonthHours: forecastNextMonth } = this.findHours(assignment.forecast);
            const { lastMonthHours: deliveredLastMonth,
                thisMonthHours: deliveredThisMonth,
                nextMonthHours: deliveredNextMonth } = this.findHours(assignment.delivered);

            result.consultants.push({
                consultantName: consultant.name,
                consultantLocation: consultant.location,
                role: assignment.role,
                forecastThisMonth: forecastThisMonth,
                forecastNextMonth: forecastNextMonth,
                deliveredLastMonth: deliveredLastMonth,
                deliveredThisMonth: deliveredThisMonth
            });

            result.forecastThisMonth += forecastThisMonth;
            result.forecastNextMonth += forecastNextMonth;
            result.deliveredLastMonth += deliveredLastMonth;
            result.deliveredThisMonth += deliveredThisMonth;

        }
        return result;
    }

    // Extract this and next month's hours from an array of HoursEntry
    private findHours(hours: HoursEntry[]): { lastMonthHours: number, thisMonthHours: number, nextMonthHours: number } {
        const now = new Date();
        const thisMonth = now.getMonth();
        const thisYear = now.getFullYear();

        const lastMonth = thisMonth === 0 ? 11 : thisMonth - 1;
        const lastYear = thisMonth === 0 ? thisYear - 1 : thisYear;

        const nextMonth = thisMonth === 11 ? 0 : thisMonth + 1;
        const nextYear = thisMonth === 11 ? thisYear + 1 : thisYear;

        const result = {
            lastMonthHours: hours.find((h) => h.month === lastMonth + 1 && h.year === lastYear)?.hours || 0,
            thisMonthHours: hours.find((h) => h.month === thisMonth + 1 && h.year === thisYear)?.hours || 0,
            nextMonthHours: hours.find((h) => h.month === nextMonth + 1 && h.year === nextYear)?.hours || 0
        };
        return result;
    }

    async addConsultantToProject(projectName: string, consultantName: string, role: string, hours: number): Promise<ApiAddConsultantToProjectResponse> {
        let projects = await this.getApiProjects(projectName, "");
        let consultants = await ConsultantApiService.getApiConsultants(consultantName, "", "", "", "", "");

        if (projects.length === 0) {
            throw new HttpError(404, `Project not found: ${projectName}`);
        } else if (projects.length > 1) {
            throw new HttpError(406, `Multiple projects found with the name: ${projectName}`);
        } else if (consultants.length === 0) {
            throw new HttpError(404, `Consultant not found: ${consultantName}`);
        } else if (consultants.length > 1) {
            throw new HttpError(406, `Multiple consultants found with the name: ${consultantName}`);
        }
        const project = projects[0];
        const consultant = consultants[0];

        // Always charge to the current month
        const remainingForecast = await AssignmentDbService.addConsultantToProject(project.id, consultant.id, role, hours);
        const message = `Added consultant ${consultant.name} to ${project.clientName} on project "${project.name}" with ${remainingForecast} hours forecast this month.`;

        return {
            clientName: project.clientName,
            projectName: project.name,
            consultantName: consultants[0].name,
            remainingForecast,
            message
        }
    }
}

export default new ProjectApiService();

File: samples/cext-trey-research-python/services/project_api_service.py
from model.base_model import Project, HoursEntry, Assignment
from model.api_model import ApiProject, ApiAddConsultantToProjectResponse
from services.project_db_service import project_db_service
from services.assignment_db_service import assignment_db_service
from services.consultant_db_service import consultant_db_service
from services.consultant_api_service import consultant_api_service
from .utilities import HttpError
from typing import List, Tuple
from datetime import datetime
import logging

class ProjectApiService:

    async def get_api_project_by_id(self, project_id: str) -> ApiProject:
        project = await project_db_service.get_project_by_id(project_id)
        assignments = await assignment_db_service.get_assignments()

        result = await self.get_api_project(project, assignments)
        return result

    async def get_api_projects(self, project_or_client_name: str, consultant_name: str) -> List[ApiProject]:
        projects = await project_db_service.get_projects()
        assignments = await assignment_db_service.get_assignments()

        # Filter on base properties
        if project_or_client_name:
            projects = [p for p in projects if (
                project_or_client_name.lower() in (p.name or '').lower() or 
                project_or_client_name.lower() in (p.clientName or '').lower()
            )]

        result = []

        # Remove duplicates
        projects = list({p.id: p for p in projects}.values())

        # Augment the base properties with assignment information
        for project in projects:
            api_project = await self.get_api_project(project, assignments)
            result.append(api_project)

        # Filter on augmented properties
        if result and consultant_name:
            result = [
                p for p in result if any(
                    consultant_name.lower() in c['consultantName'].lower() for c in p.consultants
                )
            ]

        return result

    async def get_api_project(self, project: Project, assignments: List[Assignment]) -> ApiProject:
        result = ApiProject(
        id=project.id,
        name=project.name,
        description=project.description,
        location=project.location,
        clientName=project.clientName,
        clientContact=project.clientContact,
        clientEmail=project.clientEmail,
        mapUrl=project.mapUrl,
        consultants=[],  
        forecastThisMonth=0,
        forecastNextMonth=0,
        deliveredLastMonth=0,
        deliveredThisMonth=0
        )

        # Filter the assignments related to this project
        filtered_assignments = [a for a in assignments if a.projectId == project.id]

        # Populate consultants and forecast/delivery data
        for assignment in filtered_assignments:
            # Get the consultant associated with the assignment
            consultant = await consultant_db_service.get_consultant_by_id(assignment.consultantId)

            # Calculate the forecast and delivered hours
            forecast_last_month, forecast_this_month, forecast_next_month = self.find_hours(assignment.forecast)
            delivered_last_month, delivered_this_month, _ = self.find_hours(assignment.delivered)

            # Append consultant details to the project
            result.consultants.append({
                'consultantName': consultant.get('name'),
                'consultantLocation': consultant.get('location'),
                'role': assignment.role,
                'forecastThisMonth': forecast_this_month,
                'forecastNextMonth': forecast_next_month,
                'deliveredLastMonth': delivered_last_month,
                'deliveredThisMonth': delivered_this_month,
            })

            # Update the overall project forecasts and deliveries
            result.forecastThisMonth += forecast_this_month
            result.forecastNextMonth += forecast_next_month
            result.deliveredLastMonth += delivered_last_month
            result.deliveredThisMonth += delivered_this_month

        return result

    def find_hours(self, hours: List[HoursEntry]) -> Tuple[int, int, int]:
        now = datetime.now()
        this_month = now.month
        this_year = now.year

        last_month = 12 if this_month == 1 else this_month - 1
        last_year = this_year - 1 if this_month == 1 else this_year

        next_month = 1 if this_month == 12 else this_month + 1
        next_year = this_year + 1 if this_month == 12 else this_year

        last_month_hours = next((h.get('hours') for h in hours if h.get('month') == last_month and h.get('year') == last_year), 0)
        this_month_hours = next((h.get('hours') for h in hours if h.get('month') == this_month and h.get('year') == this_year), 0)
        next_month_hours = next((h.get('hours') for h in hours if h.get('month') == next_month and h.get('year') == next_year), 0)

        return last_month_hours, this_month_hours, next_month_hours

    async def add_consultant_to_project(self, project_name: str, consultant_name: str, role: str, hours: int) -> ApiAddConsultantToProjectResponse:
        projects = await self.get_api_projects(project_name, "")
        consultants = await consultant_api_service.get_api_consultants(consultant_name, "", "", "", "", "")

        if len(projects) == 0:
            raise HttpError(404, f"Project not found: {project_name}")
        elif len(projects) > 1:
            raise HttpError(406, f"Multiple projects found with the name: {project_name}")
        elif len(consultants) == 0:
            raise HttpError(404, f"Consultant not found: {consultant_name}")
        elif len(consultants) > 1:
            raise HttpError(406, f"Multiple consultants found with the name: {consultant_name}")

        project = projects[0]
        consultant = consultants[0]

        # Always charge to the current month
        remaining_forecast = await assignment_db_service.add_consultant_to_project(project.id, consultant.id, role, hours)
        message = f"Added consultant {consultant.name} to {project.clientName} on project '{project.name}' with {remaining_forecast} hours forecast this month."

        return ApiAddConsultantToProjectResponse(
            clientName=project.clientName,
            projectName=project.name,
            consultantName=consultant.name,
            remainingForecast=remaining_forecast,
            message=message
        )


# This would initialize the service in a similar way to the TypeScript `export default`
project_api_service = ProjectApiService()